### PR TITLE
KOGITO-4581 - Fix template exception handling and refactoring

### PR DESCRIPTION
**Pull request by r00ta**
JIRA: https://issues.redhat.com/browse/KOGITO-4581

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/master/kogito-ide-config)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

### CODE REVIEW

File: kogito-codegen-modules/kogito-codegen-decisions/src/main/java/org/kie/kogito/codegen/decision/DecisionRestResourceGenerator.java [hunk 1]
```diff
@@ -302,39 +299,6 @@ private void interpolateInputType(ClassOrInterfaceDeclaration template) {
...
-                    return parent.getNameAsString().startsWith("decisionService_");
-                })
-                .collect(Collectors.toList());
-
-        objectReturnTypes.forEach(type -> type.setName("Object"));
-
-        outputTypeOccurrences.removeAll(objectReturnTypes);
-        outputTypeOccurrences.forEach(type -> type.setName(outputType));
-    }
-
```
Comment (tarilabs) [inline kogito-codegen-modules/kogito-codegen-decisions/src/main/java/org/kie/kogito/codegen/decision/DecisionRestResourceGenerator.java:337]: Sorry, no.
I'm pretty sure if you remove this, `stronglytyped` would no longer work.

File: kogito-codegen-modules/kogito-codegen-decisions/src/main/resources/class-templates/DecisionRestResourceQuarkusTemplate.java [hunk 2]
```diff
@@ -47,7 +58,7 @@
     @org.eclipse.microprofile.openapi.annotations.responses.APIResponse(content = @org.eclipse.microprofile.openapi.annotations.media.Content(mediaType = "application/json", schema = @org.eclipse.microprofile.openapi.annotations.media.Schema(ref = "/dmnDefinitions.json#/definitions/OutputSet1")), description = "DMN output")
     @io.swagger.v3.oas.annotations.parameters.RequestBody(content = @io.swagger.v3.oas.annotations.media.Content(mediaType = "application/json",schema = @io.swagger.v3.oas.annotations.media.Schema(ref = "/dmnDefinitions.json#/definitions/InputSet1")), description = "DMN input")
     @io.swagger.v3.oas.annotations.responses.ApiResponse(content = @io.swagger.v3.oas.annotations.media.Content(mediaType = "application/json",schema = @io.swagger.v3.oas.annotations.media.Schema(ref = "/dmnDefinitions.json#/definitions/OutputSet1")), description = "DMN output")
-    public $outputType$ dmn($inputType$ variables) {
+    public Response dmn($inputType$ variables) {
```
Comment (tarilabs) [inline kogito-codegen-modules/kogito-codegen-decisions/src/main/resources/class-templates/DecisionRestResourceQuarkusTemplate.java:61]: as above.

File: kogito-codegen-modules/kogito-codegen-decisions/src/main/resources/class-templates/DecisionRestResourceSpringTemplate.java [hunk 3]
```diff
@@ -40,17 +42,25 @@
...
     @org.eclipse.microprofile.openapi.annotations.responses.APIResponse(content = @org.eclipse.microprofile.openapi.annotations.media.Content(mediaType = "application/json", schema = @org.eclipse.microprofile.openapi.annotations.media.Schema(ref = "/dmnDefinitions.json#/definitions/OutputSet1")), description = "DMN output")
     @io.swagger.v3.oas.annotations.parameters.RequestBody(content = @io.swagger.v3.oas.annotations.media.Content(mediaType = "application/json",schema = @io.swagger.v3.oas.annotations.media.Schema(ref = "/dmnDefinitions.json#/definitions/InputSet1")), description = "DMN input")
     @io.swagger.v3.oas.annotations.responses.ApiResponse(content = @io.swagger.v3.oas.annotations.media.Content(mediaType = "application/json",schema = @io.swagger.v3.oas.annotations.media.Schema(ref = "/dmnDefinitions.json#/definitions/OutputSet1")), description = "DMN output")
-    public $outputType$ dmn(@RequestBody(required = false) $inputType$ variables,
+    public ResponseEntity<?> dmn(@RequestBody(required = false) $inputType$ variables,
```
Comment (tarilabs) [inline kogito-codegen-modules/kogito-codegen-decisions/src/main/resources/class-templates/DecisionRestResourceSpringTemplate.java:63]: as above.

**Review by tarilabs — CHANGES_REQUESTED** [review]
No sorry this is not convincing, I'm pretty sure you are eviscerating `stronglytyped`.

**Comment by r00ta** [comment]
> No sorry this is not convincing, I'm pretty sure you are eviscerating `stronglytyped`.

As clarified by call, switching to `Response` and `ResponseEntity` is not going to break `stronglytyped`. With these changes we remove a bit of complexity and codegen, which is good. 
If anybody sees any problem with these changes, let me know

File: kogito-codegen-modules/kogito-codegen-decisions/src/main/java/org/kie/kogito/codegen/decision/DecisionRestResourceGenerator.java (hunk 1 above)
Comment (r00ta) [reply kogito-codegen-modules/kogito-codegen-decisions/src/main/java/org/kie/kogito/codegen/decision/DecisionRestResourceGenerator.java:337]: See main comment

**Review by tarilabs — APPROVED** [review]
as anticipated within call of https://github.com/kiegroup/kogito-runtimes/pull/1113#issuecomment-790846234
I re-checked with @danielezonca 

It is fine this modification since:
 - we have a orthogonal feature which would render the OAS return information anyway (openapi descriptors)
 - on the long term, for "programmatic API / scaffolding" perspective, it is fine the return type is a bit more opaque (Response instead of Object) so dealing with a REST semantic (Response); if a end-user wants to bind to the java method anyway, it can always make a call to `body` and extract the return payload. Tho it would be more appropriate to get a local client call.

So it is fine to be departing from the previous "requirement" to be "kogito-ergo-domain" in the return type of the java method for the REST call, approving.

**Comment by r00ta** [comment]
Hi @hellowdan , let me know if you can review this or if another QE should look at it. Thanks!

**Review by danielezonca — APPROVED** [review]

**Review by jiripetrlik — COMMENTED** [review]
@r00ta 
Would it be possible to add some tests which checks that exceptions are handled correctly and correct exception metrics are provided?

**Comment by r00ta** [comment]
> @r00ta
> Would it be possible to add some tests which checks that exceptions are handled correctly and correct exception metrics are provided?

Hi @jiripetrlik , I've created this ticket. I'll start working on it as soon as I'm done with other tasks https://issues.redhat.com/browse/KOGITO-4624

**Review by jiripetrlik — APPROVED** [review]
@r00ta 
Thank you for ticket for additional test coverage! Looks good to me.

**Pull request merged**